Source/WebCore/page/Navigation.cpp

```diff
@@ -442,6 +460,40 @@ void Navigation::updateForNavigation(Ref<HistoryItem>&& item, NavigationNavigati
         disposedEntry->dispatchEvent(Event::create(eventNames().disposeEvent, { }));
 }
 
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#update-the-navigation-api-entries-for-reactivation
+void Navigation::updateForReactivation(Vector<Ref<HistoryItem>>& newHistoryItems, HistoryItem& reactivatedItem)
+{
+    if (hasEntriesAndEventsDisabled())
+        return;
+
+    Vector<Ref<NavigationHistoryEntry>> newEntries;
+    Vector<Ref<NavigationHistoryEntry>> oldEntries = WTFMove(m_entries);
+
+    for (Ref item : newHistoryItems) {
+        RefPtr<NavigationHistoryEntry> newEntry;
+
+        for (size_t entryIndex = 0; entryIndex < oldEntries.size(); entryIndex++) {
+            auto& entry = oldEntries.at(entryIndex);
+            if (entry->associatedHistoryItem() == item) {
+                newEntry = entry.ptr();
+                oldEntries.remove(entryIndex);
+                break;
+            }
+        }
+
+        if (!newEntry)
+            newEntry = NavigationHistoryEntry::create(scriptExecutionContext(), WTFMove(item));
+
+        newEntries.append(newEntry.releaseNonNull());
```

Comment:
Do we want to append to the end if we removed from the middle?

Reply:
This is in a loop creating a new list so it adds everything that had a shared history linearly.

The final step below sets the active entry (`m_currentEntryIndex = getEntryIndexOfHistoryItem(m_entries, reactivatedItem);`